crates/profiling/src/vtune_linux.rs

```diff
@@ -0,0 +1,138 @@
+//! Adds support for profiling jitted code using VTune Amplifier
+//!
+//! Build:
+//!     cargo build --features=vtune
+//! Profile:
+//!     amplxe-cl -run-pass-thru=--no-altstack -v -collect hotspots target/debug/wasmtime --vtune test.wasm
+//!
+//! Note: amplxe-cl is a command-line tool for Vtune which should be installed.
+
+use anyhow::Result;
+use core::ptr;
+use ittapi_rs::*;
+use std::boxed::*;
+use std::collections::HashMap;
+use std::ffi::CString;
+use std::sync::Mutex;
+use wasmtime_environ::entity::PrimaryMap;
+use wasmtime_environ::wasm::DefinedFuncIndex;
+use wasmtime_environ::Module;
+use wasmtime_runtime::VMFunctionBody;
+
+/// Interface for driving the ittapi for VTune support
+pub struct VTuneAgent {
+    // Note that we use a mutex internally to serialize state updates
+    // since multiple threads may be sharing this agent.
+    state: Mutex<State>,
+}
+
+/// Interface for driving vtune
+#[derive(Clone, Debug, Default)]
+struct State {
+    /// Unique identifier for the jitted function
+    method_id: HashMap<(usize, DefinedFuncIndex), u32>,
+}
+
+impl VTuneAgent {
+    /// Intialize a VTuneAgent and write out the header
+    pub fn new() -> Result<Self> {
+        let state = State {
+            method_id: HashMap::new(),
+        };
+        Ok(VTuneAgent {
+            state: Mutex::new(state),
+        })
+    }
+}
+
+impl State {
+    /// Return the unique method ID for use with the ittapi
+    pub fn get_method_id(&mut self, module_id: usize, func_idx: DefinedFuncIndex) -> u32 {
+        let method_id;
+        unsafe {
+            method_id = self
+                .method_id
+                .entry((module_id, func_idx))
+                .or_insert_with(|| iJIT_GetNewMethodID());
```

Comment:
How come this is cached internally? I don't think we publish code for the same module twice, so wouldn't we want to always create a new method id?

Reply:
This is a good question. I saved this for supporting other calls that would rely on the same methodID. Right now this pr only calls method_load for a function and then moves on, but there are other events that can be recorded that aren't included with this particular patch (https://software.intel.com/en-us/vtune-help-ijit-notifyevent). Also .. I wasn't sure what the invoke flag was allowing now. In the past I ran an example wondering why my codes were running twice, only to realize I was using invoke when I did not need to. This was some action.rs file where I had extra hooks, which looks like it is gone now but I was calling profiling hooks twice when maybe it wasn't needed. I wanted to cache methodID to prepare for this possibility as well.

Also note, this instrumentation is targeting jitted code but future updates should include APIs for help in instrumenting wasmtime itself: https://software.intel.com/en-us/vtune-help-instrumenting-your-application#2CCDFEFA-30CD-4899-A86B-54C9F873C823. This is different from recording the methodID, but the same VTune agent interface here I expected would be used to cash domain ids and other handles that help with marking regions.

Reply:
Ah ok caching makes sense, but I think that here it may make more sense to insert it into the map and assert that it was the first time we saw it? Modules should only be reigstered once and this would allow other VTune methods to get called as well.

Reply:
👍